src/util/AsioHelpers.h

```diff
@@ -103,6 +107,71 @@ auto runFunctionOnExecutor(Executor executor, Function function,
   return net::async_initiate<CompletionToken, Signature>(
       std::move(initiatingFunction), completionToken);
 }
+
+/// Helper function, that checks a cancellation handle regularly while waiting
+/// on an awaitable object that doesn't do this on its own. It's always better
+/// to integrate cancellation checks right into the awaitable itself, but
+/// sometimes this doesn't work because it's part of a library or boost asio
+/// itself. Once `timerRunning` resolves to `false` (or the awaitable is
+/// finished, whatever happens first), the cancellation checks are stopped.
+template <typename T>
```

Comment:
For this abstraction it is important, that the executor of the awaitable is single_threaded (strand or only one thread accessing an `io_context` or threadpool, otherwise you again have the data races in the `awaitable operator&&` .
This should be at least stated somewhere here.